### DISCUSSION TITLE
Use AsyncConnectionWrapper with TLS connection

### DISCUSSION THREAD

**Discussion by DevLazio** [weiznich/diesel_async#105] [Q&A] [answered]
Hello,

I have been trying to use the newly release AsyncConnectionWrapper to run my migrations at the start of my program.
I may have missed something but it doesn't seem to be a way to get an AsyncConnectionWrapper with TLS capabilities : the only documented way to get an AsyncConnectionWrapper is through the establish() function, which only accept an URL and no config.
Predictably, the connection fail with "error performing TLS handshake: no TLS implementation configured".

Is there a way to add a TLS configuration or to wrap around a previously setup AsyncDieselConnectionManager ?

**Comment by weiznich** [comment, marked as answer]
To cite from the [release announcement post](https://blog.weiznich.de/blog/diesel-async-0-4/):

> Moreover, there are differences between the provided connection wrapper and existing diesel connection types in terms of functionality and behavior. These discrepancies include variations in handling database URL's, TLS configurations, and row streaming from the database. As a result the new connection type is not supposed to be a drop in replacement for existing diesel connection types.

So yes that's currently "expected" behavior and there is no other way to configure the TLS setup. 
That written: I'm happy to accept PR's to add this feature, but I likely won't have the time to address this soon due to other issues. After all that's a request for a "new" feature, and I consider fixing existing bugs more important for now.

**Reply by weiznich** [reply]
To add some idea for a possibly accepted implementation here:

We could just add a generic `TryFrom` implementation for all async connections + all `BlockOn` types that implement `Default`.

**Reply by DevLazio** [reply]
Why not a simple `From` like that:
```rust
impl<C, B> From<C> for AsyncConnectionWrapper<C, B>
        where
            C: crate::AsyncConnection,
            B: BlockOn + Send,
    {
        fn from(inner: C) -> Self {
            Self {
                inner,
                runtime: B::get_runtime(),
            }
        }
    }
```
it seems to work fine for me.

**Reply by weiznich** [reply]
That would be also fine.
